### PR TITLE
feat(policies): evaluate policies on attestation push command

### LINKED ISSUE #122: Support policies for attestations
Currently, no verification is done when an attestation is received in the control plane. Creation and validation checks are done in the client side, but not on the server side. 

We should allow operators to attach `rego` or potentially `cue` policies to their contracts and these should get evaluated during the reception of the attestation.  

We created a task #35 which will get superseded by this functionality since 35 is in fact a policy check that some materials exist and that the runner type is correct.

re: implementation

- We can start with making this check an enforcement, meaning that if the policy check fails the `att push` command will too since the check for the first version could be sync.
- We could implement this check as a fan-out integration. Create it by default in your org and automatically attach it to each of your workflows.

About the policy format.

We should probably aim towards Open Policy Agent (OPA) and leverage (conftest) logic. We should take a look at policy-controller [way of doing this](https://docs.sigstore.dev/policy-controller/overview/#configuring-policy-that-validates-attestations)

On the UX side of things we could allow attaching a policy to our contract today.

example from policy-controller that could map our current contract too.

using `cue` 
```yaml
policy:
        type: cue
        data: |
          predicateType: "https://cosign.sigstore.dev/attestation/v1"
          predicate: "foobar e2e test"
```

or rego

```yaml
policy:
        type: rego
        data: |
          package sigstore
          default isCompliant = false
          isCompliant {
            input.predicateType == "https://cosign.sigstore.dev/attestation/v1"
            input.predicate == "foobar e2e test"
          }
```

Note: If we go ahead with the implementation based on an integration. We would need to extend the integrations model to read from its result and apply it to the workflow runs. Currently they are just fire and forget.

```[tasklist]
### Tasks
- [x] Propose policy API (protos)
```

### PR DESCRIPTION
**Pull request by jiparis**
This PR is a follow up that connects all the pieces to evaluate any policy present in the schema before pushing the attestation. The PR includes:
```[tasklist]
### features
- [x] Evaluating policies against materials during `att add`
- [x] Evaluating policies against statement (not signed) during `att push`
- [x] validate schema in crafter (during `contract create` and `contract update`)
- [x] Storing evaluation results in the predicate, included in the signed attestation, and sent to the controlplane.
``` 

Policies can be specified and linked to any schema. For example, this policy can be used to evaluate any material of type SBOM_SPDX_JSON, and check that was created with Syft.
```
# File policy-made-with-syft.yaml
apiVersion: workflowcontract.chainloop.dev/v1
kind: Policy
metadata:
  name: made-with-syft
spec:
  type: SBOM_SPDX_JSON
  embedded: |
    package main

    import future.keywords.in

    deny[msg] {
      not made_with_syft

      msg := "Not made with syft"
    }

    made_with_syft {
      some creator in input.creationInfo.creators
      contains(creator, "syft")
    }
```
Policy code can also be detached and referenced from the spec. I'll update the documentation in a separate PR.
Then, it can be linked to any contract:
```
# my-schema.yaml
schemaVersion: v1
materials:
  - name: sbom
    type: SBOM_SPDX_JSON
policies:
  materials:
    - ref: "policy-made-with-syft.yaml" 
  attestation:
    - ref: "policy-author.yaml"
```
Note the `policies.materials` and `policies.attestation` section, denoting policies to be applied to all materials or the whole attestation (before signing) accordingly. In the case of materials, since the policy has a `type` in the specification, it will be applied to all materials that matches that type. There are other options and filters to flexibilize the policy selection criteria, allowing to apply policies to specific materials. For example, it can be filtered by name:
```
...
policies:
  materials:
    - ref: "policy-made-with-syft.yaml" 
      selector:
         name: "sbom" # only apply the policy to the "sbom" material
```
If there is no `type` in the policy spec, it will be applied regardless the material has a compatible type or not. This is useful for custom validations and artifacts.

Closes #122

### CODE REVIEW

**Review by javirln — APPROVED** [review]
Awesome!!

File: internal/attestation/renderer/chainloop/chainloop.go
Comment (jiparis) [inline internal/attestation/renderer/chainloop/chainloop.go:81]: Moved outside the common section.

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:175]: I believe we have some code that does this already, so we could just pass the connection ready to the action?

Comment (migmartri) [reply pkg/policies/policies.go:175]: https://github.com/migmartri/chainloop/blob/40f96e6edc5c4f3bba347b30b44af88614dd8e30/app/cli/cmd/artifact.go#L38-L62

File: app/cli/cmd/attestation_push.go
Comment (migmartri) [inline app/cli/cmd/attestation_push.go:69]: Why don't we just pass the CAS connection? We have already code in the CMD package to generate the connection https://github.com/migmartri/chainloop/blob/40f96e6edc5c4f3bba347b30b44af88614dd8e30/app/cli/cmd/artifact.go#L38-L62

you can see an example in artifact_download command

File: app/cli/internal/action/attestation_add.go
Comment (migmartri) [inline app/cli/internal/action/attestation_add.go:23]: Are you using any specific sorting plugin?

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:146]: if we happen to do this validation in `attestation add` we would not need to download the file.

Downloading at push when the file is still there doesn't seem optimal and potentially problematic with big files.

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:112]: To me the name should be a filter between all the types that a policy affects to, not the differentiator between a material policy or an attestation policy. WDYT?

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:142]: are you handling when the materials are stored inline?

**Review by migmartri — COMMENTED** [review]
Thanks @jiparis for the PR, looking promising. I've not finished the review yet but I want to leave some early thoughts.

- I am not convinced that running the material policies on push is a good idea, especially taking into account that we download the materials again. Why can't we run the policies on `att add` and store the violation in the crafting state?
- we should revisit the use of `name` in the attachment, the way it's implemented today will not allow it working with contractless materials. To me, the `kind` is in the policy and name would be an optional selector to scope down the result.

Let me know you thoughts, happy to chat tomorrow :)

File: internal/attestation/crafter/crafter_test.go
Comment (migmartri) [inline internal/attestation/crafter/crafter_test.go:239]: should this test somehow check the result? `want`?

File: internal/attestation/renderer/chainloop/chainloop.go
Comment (migmartri) [inline internal/attestation/renderer/chainloop/chainloop.go:168]: out of curiosity, why replacing the pointers?

File: internal/attestation/renderer/chainloop/v02.go
Comment (migmartri) [inline internal/attestation/renderer/chainloop/v02.go:183]: could be worse, I think you did pretty well here :)

**Review by migmartri — COMMENTED** [review]
more feedback

File: internal/attestation/renderer/chainloop/v02.go
Comment (migmartri) [inline internal/attestation/renderer/chainloop/v02.go:178]: do you need to store the result as structpb? I thought we had a custom predicate that we were mapping ourselves?

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:100]: we should probably also store when a policy was successfully ran not only when it's a violation, what do you think?

File: app/cli/internal/action/attestation_add.go
Comment (jiparis) [reply app/cli/internal/action/attestation_add.go:23]: Just `goimports --local`

File: app/cli/cmd/attestation_push.go
Comment (jiparis) [reply app/cli/cmd/attestation_push.go:69]: We can't, because CAS connection is specific for a given digest, and here we wanted to download "any" material, so we need to create specific connections for each.

File: app/controlplane/api/workflowcontract/v1/crafting_schema.proto
Comment (jiparis) [reply app/controlplane/api/workflowcontract/v1/crafting_schema.proto:152]: To avoid import cycles. In particular `policy.proto` needed material types, and crafting_schema needed policies. We could extract the common part to a third proto, but that would require several unrelated changes throughout all the code.

File: internal/attestation/renderer/chainloop/v02.go
Comment (jiparis) [reply internal/attestation/renderer/chainloop/v02.go:178]: Just because we have some `oneof`s in the policies, which is incompatible with structpb conversion.

File: pkg/policies/policies.go
Comment (jiparis) [reply pkg/policies/policies.go:100]: In that case `Violations` will be empty, but the policy will still be there.

File: pkg/policies/policies.go
Comment (jiparis) [reply pkg/policies/policies.go:112]: I like it. We'll give priority to the type, and then filter by name if it's provided. I'll send another revision of the PR.

File: pkg/policies/policies.go
Comment (jiparis) [reply pkg/policies/policies.go:142]: Yes, in the policies logic, it checks if the material is inline, and in that case, it retrieves the content from the `Content` field.

File: pkg/policies/policies.go
Comment (jiparis) [reply pkg/policies/policies.go:146]: Yes, I agree.

File: pkg/policies/policies.go
Comment (jiparis) [reply pkg/policies/policies.go:175]: It needs the digest, which depends on each material. We might want to revisit that restriction, but currently we must create a new client for each.

**Comment by jiparis** [comment]
Please put reviews on hold, I'll send another revision with requested changes.

File: internal/attestation/renderer/chainloop/chainloop.go
Comment (jiparis) [reply internal/attestation/renderer/chainloop/chainloop.go:168]: I think at some point during my changes it needed to be preallocated. Not sure if that's still the case.

File: internal/attestation/crafter/materials/materials.go
Comment (jiparis) [inline internal/attestation/crafter/materials/materials.go:215]: I've moved this to avoid import loops.

File: app/cli/cmd/attestation_push.go
Comment (migmartri) [inline app/cli/cmd/attestation_push.go:26]: These changes make the PR very noisy, you mentioned that you use `goimports` but I can not reproduce the change locally with latest version of goimports.

Are you sure you do not have some additional configuration in the editor?

```
# From main
goimports -w app/cli/cmd/attestation_push.go 
git status
[empty]
```

I think we need to figure this out before we merge, otherwise we'll be doing back and forth continuously

Comment (jiparis) [reply app/cli/cmd/attestation_push.go:26]: Ok, let me undo these accidental changes, since they are not related.

File: internal/attestation/crafter/crafter_test.go
Comment (jiparis) [reply internal/attestation/crafter/crafter_test.go:251]: Done!

File: internal/attestation/crafter/testdata/contracts/with_policy_embedded.yaml
Comment (migmartri) [inline internal/attestation/crafter/testdata/contracts/with_policy_embedded.yaml:4]: is this embedded?

File: internal/attestation/crafter/testdata/contracts/with_rego.yaml
Comment (migmartri) [inline internal/attestation/crafter/testdata/contracts/with_rego.yaml:3]: in the future we should also have some examples with `material` attachments, all of them in this directory seem to be attestation.

File: internal/attestation/crafter/crafter.go
Comment (migmartri) [inline internal/attestation/crafter/crafter.go:188]: Isn't this embedding the policy which we said we didn't want to do?

File: internal/attestation/crafter/crafter.go
Comment (migmartri) [inline internal/attestation/crafter/crafter.go:577]: This is very clear now, thanks :)

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:47]: does this comment make sense anymore? I can't see any reference to engine

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:57]: nitpick, add empty line

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:53]: I have a problem with the naming here, it seems that the term `Policy` is being overused for both the actual policy and for evaluations of the policy on a material for example.

Wouldn't make sense to change the name of this one to `PolicyEvaluation`?

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:188]: Could you add a comment on what this method does? I mean, the different logic paths it goes through.

In another patch, it would be great to test this method so it's used as documentation too.

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:163]: nitpick. I find easier to read separated statements

```
if artifactPath == "" {
return nil, errors...
}

```
rawMaterial, err = os.ReadFile(artifactPath)
		if err != nil {
			return nil, fmt.Errorf("failed to read material content: %w", err)
		}
```

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:173]: Good catch

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:179]: can you double check that we do not have a method that does this already? I remember creating a common method for marshalling/unmarshalling from JSON to Statement

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:81]: nitpick, instead of `res` call it violations?

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:242]: I like it! 

Some testing later on would be ideal

File: internal/attestation/renderer/chainloop/v02.go
Comment (migmartri) [inline internal/attestation/renderer/chainloop/v02.go:41]: as I mentioned, `Policy` might not be a good name here. PolicyEval?

File: internal/attestation/renderer/renderer.go
Comment (migmartri) [inline internal/attestation/renderer/renderer.go:155]: I'd add a comment explaining why this method needs to leverage structpb and so on

**Review by migmartri — APPROVED** [review]
Very nice!

Approving modulo making sure that we are not embedding the policies on contract create/update

File: internal/attestation/crafter/crafter.go
Comment (jiparis) [reply internal/attestation/crafter/crafter.go:188]: I removed the embedding logic, but I should change the method name 😞 . Let me change it.

File: internal/attestation/crafter/testdata/contracts/policy_embedded.yaml
Comment (jiparis) [reply internal/attestation/crafter/testdata/contracts/policy_embedded.yaml:2]: makes sense, I'll do it.

File: pkg/policies/policies.go
Comment (jiparis) [reply pkg/policies/policies.go:53]: Makes all sense. I'll create a specific struct for this.

File: app/controlplane/api/workflowcontract/v1/crafting_schema.proto
Comment (jiparis) [reply app/controlplane/api/workflowcontract/v1/crafting_schema.proto:192]: Right, it's not yet implemented.

File: internal/attestation/crafter/testdata/contracts/with_policy_embedded.yaml
Comment (jiparis) [reply internal/attestation/crafter/testdata/contracts/with_policy_embedded.yaml]: The code in the policy, not the policy in the contract.

File: pkg/policies/policies.go
Comment (jiparis) [reply pkg/policies/policies.go:187]: Yes, it's tested thoroughly in policies_test.go. I'll add the comment.

File: internal/attestation/renderer/renderer.go
Comment (jotadrilo) [inline internal/attestation/renderer/renderer.go:177]: You should use the string representation of the MaterialType enum to avoid future inconsistencies

```suggestion
	p.PolicyEvaluations[schemaapi.CraftingSchema_Material_ATTESTATION.String()] = policyResults
```

Comment (jiparis) [reply internal/attestation/renderer/renderer.go:177]: Ey, thanks for the suggestion @jotadrilo . But in this case it's not a material, but policies applied to the whole attestation. I know the naming is confusing, we might choose a different key for this particular case. I'll give it a try.

File: pkg/policies/policies.go
Comment (migmartri) [inline pkg/policies/policies.go:180]: told you :)

File: internal/attestation/renderer/renderer.go
Comment (jotadrilo) [reply internal/attestation/renderer/renderer.go:177]: But you are mapping materials and policies, right? In other groups, you are grouping policies by their material name.

If this specific grouping has a different semantic, I'd better use a different field in the in-toto predicate (v2), or add a reserved MaterialType name for this kind of grouping.

File: pkg/policies/policies.go
Comment (jiparis) [reply pkg/policies/policies.go:180]: Yes! And it also implements URLEncoding which is is supported by DSSE in addition to StdEncoding.

File: internal/attestation/renderer/renderer.go
Comment (jiparis) [reply internal/attestation/renderer/renderer.go]: Yes, we were evaluating those options. We'll revisit this in next iterations, as it might be part of a batch with other improvements. Thanks!

**Pull request merged**
